src/coreclr/jit/codegenlinear.cpp

```diff
@@ -721,9 +721,7 @@ void CodeGen::genCodeForBBlist()
 
                     if ((call != nullptr) && (call->gtOper == GT_CALL))
                     {
-                        if ((call->AsCall()->gtCallMoreFlags & GTF_CALL_M_DOES_NOT_RETURN) != 0 ||
-                            ((call->AsCall()->gtCallType == CT_HELPER) &&
-                             Compiler::s_helperCallProperties.AlwaysThrow(call->AsCall()->GetHelperNum())))
+                        if (call->AsCall()->IsNoReturn())
```

Comment:
Not related to this change, but I was wondering - why do we do `(call->gtOper == GT_CALL)`? 
Should it be possible at this point for a BBJ_THROW to end with anything other than a call?

Reply:
I am mostly concerned about a possibility that the throwing call is somewhere inside the block and followed by some unreachable junk. Then we either emit the junk unnecessarily, or, worse, emitter may skip/optimize it and we end up with the same state we are trying to prevent here.  
So I wonder - is there `GT_SOMETHING` that is not a call, and yet it throws and can conclude a `BBJ_THROW`?

Reply:
I tested this out by asserting the last node is a call if the block has any IR, and the assert did hit. Here's some example IR it hit for:
```
------------ BB04 [0003] [024..02A) (throw), preds={BB03} succs={}
N069 (???,???) [000043] -----------                            IL_OFFSET void   INLRT @ 0x024[E-] REG NA
N071 (  3, 10) [000046] Hc---------                   t46 =    CNS_INT(h) long   0x7ffc04235008 ftn REG NA
                                                            /--*  t46    long   
N073 (  5, 12) [000047] nc--G------                   t47 = *  IND       long   REG NA
                                                            /--*  t47    long   control expr
N075 ( 14,  5) [000020] --CXG------                   t20 = *  CALL      int    Microsoft.FSharp.Core.Operators+OperatorIntrinsics:alreadyFinished[int]():int REG rax $c4
                                                            /--*  t20    int    
N077 ( 15,  6) [000022] --CXG------                         *  RETURN    int    REG NA $VN.Void
```

That call is marked as not returning, but there's still a `GT_RETURN` after it; this seems to confirm your concerns. Perhaps we need a dead code elimination check somewhere that deletes any IR after no-return calls, though a quick-and-dirty fix might be to always emit the breakpoint after no-return calls (regardless of whether they're the last node or not) in `CodeGen::genCall`, or somewhere similar.

Reply:
Morph has an exemption for tail calls that are no return, and will not prune the IR afterwards. So likely the method above is a tail call? Also there is this note:

https://github.com/dotnet/runtime/blob/543b2cc740e3e323244c7c4e7a53dc954abf6826/src/coreclr/jit/morph.cpp#L7825-L7844

which the IR dump above seems to contradict, since it looks from the dump that the block kind is BBJ_THROW. I wonder if the comment is now stale?

Earlier on morph will reject implicit tail calls to no return methods, I wonder if we should just do the same for explicit tail calls, even though technically we'd be violating the intent of the `.tail` prefix. Then we would not have these special cases.

> a quick-and-dirty fix might be to always emit the breakpoint after no-return calls

There's also a concern that with profiler-driven rejitting a method that might always throw when the JIT looked at it could be updated to not always throw. I don't know if there is any good reason to support this sort of rejitting, but we should make sure there's an obvious failure if someone does it since the jitted code may depend on the old behavior. So I think in general we will want to put a breakpoint afterwards.

Reply:
> Morph has an exemption for tail calls that are no return, and will not prune the IR afterwards. So likely the method above is a tail call?

In the above case, the call is an implicit tail call to a no-return method, so `fgMorphPotentialTailCall` rejected it, and `fgMorphCall` set `fgRemoveRestOfBlock = true`. The `GT_CALL` node is an operand of a `GT_RETURN` node, which is in the last statement of the block, so `fgRemoveRestOfBlock` didn't actually remove any IR, but it still converted the block into a `BBJ_THROW`. Perhaps we could add a check that prevents converting the block into a `BBJ_THROW` if it ends with a `GT_RETURN` node, so that we can always expect the last node in a `BBJ_THROW` to be a `GT_CALL`? I'm not sure if we lose any benefits by doing this -- maybe should still mark the block as rarely run?